presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxByNAggregation.java

```diff
@@ -0,0 +1,177 @@
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.operator.aggregation.state.MinMaxByNState;
+import com.facebook.presto.operator.aggregation.state.MinMaxByNStateFactory;
+import com.facebook.presto.operator.aggregation.state.MinMaxByNStateSerializer;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.type.ArrayType;
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Ints;
+
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static com.facebook.presto.metadata.Signature.orderableTypeParameter;
+import static com.facebook.presto.metadata.Signature.typeParameter;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INDEX;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.NULLABLE_BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.STATE;
+import static com.facebook.presto.operator.aggregation.AggregationUtils.generateAggregationName;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractMinMaxByNAggregation
+        extends ParametricAggregation
+{
+    private static final MethodHandle INPUT_FUNCTION = methodHandle(AbstractMinMaxByNAggregation.class, "input", BlockComparator.class, Type.class, Type.class, MinMaxByNState.class, Block.class, Block.class, int.class, long.class);
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(AbstractMinMaxByNAggregation.class, "combine", MinMaxByNState.class, MinMaxByNState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(AbstractMinMaxByNAggregation.class, "output", ArrayType.class, MinMaxByNState.class, BlockBuilder.class);
+
+    private final String name;
+    private final Function<Type, BlockComparator> typeToComparator;
+    private final Signature signature;
+
+    protected AbstractMinMaxByNAggregation(String name, Function<Type, BlockComparator> typeToComparator)
+    {
+        this.name = requireNonNull(name);
```

Comment:
use `requireNonNull(name, "name is null")`. I made an IntelliJ macro to insert this, which makes it easy to add